content/advent-2018/llvm-ir-and-go.md

```diff
@@ -0,0 +1,311 @@
+
+<img alt="phi instruction" src="/postimages/advent-2018/llvm-ir-and-go/phi_instruction.png" width="300">
+
+## LLVM IR library in pure Go
+
+The two main libraries for working with LLVM IR in Go are:
+
+* [llvm.org/llvm/bindings/go/llvm](https://llvm.org/svn/llvm-project/llvm/trunk/bindings/go/README.txt): the official LLVM bindings for the Go programming language.
+* [github.com/llir/llvm](https://github.com/llir/llvm): a pure Go library for interacting with LLVM IR.
+
+The official LLVM bindings for Go uses Cgo to provide access to the rich and powerful API of the LLVM compiler framework, while the `llir/llvm` project is entirely written in Go and relies on LLVM IR to interact with the LLVM compiler framework.
+
+This post focuses on `llir/llvm`, but should generalize to working with other libraries as well.
+
+### Why write a new library?
+
+The primary motivation for developing a pure Go library for interacting with LLVM IR was to make it more fun to code compilers and static analysis tools that rely on and interact with the LLVM compiler framework. In part because the compile time of projects relying on the official LLVM bindings for Go could be quite substantial (Thanks to [@aykevl](https://github.com/aykevl), the author of [TinyGo](https://github.com/aykevl/tinygo), there are now ways to speed up the compile time by dynamically linking against a system-installed version of LLVM[^4]).
+
+Another leading motivation was to try and design an idiomatic Go API from the ground up. The main difference between the API of the LLVM bindings for Go and `llir/llvm` is how LLVM values are modelled. In the LLVM bindings for Go, LLVM values are modelled as [a concrete struct type](https://godoc.org/llvm.org/llvm/bindings/go/llvm#Value), which essentially contains every possible method of every possible LLVM value. My personal experience with using this API is that it was difficult to know what subsets of methods you were allowed to invoke for a given value. For instance, to retreive the Opcode of an instruction, you'd invoke the [InstructionOpcode](https://godoc.org/llvm.org/llvm/bindings/go/llvm#Value.InstructionOpcode) method -- which is quite intuitive. However, if you happen to invoke the [Opcode](https://godoc.org/llvm.org/llvm/bindings/go/llvm#Value.Opcode) method instead (which is used to retreive the Opcode of constant expressions), you'd get the runtime errors _"cast&lt;Ty&gt;() argument of incompatible type!"_.
+
+The `llir/llvm` library was therefore designed to provide compile time guarantes by further relying on the Go type system. LLVM values in `llir/llvm` are modelled as [an interface type](https://godoc.org/github.com/llir/llvm/ir/value#Value). This approach only exposes the minimum set of methods shared by all values, and if you want to access more specific methods or fields, you'd use a type switch (as illustrated in the [analysis example](#analysis-example-processing-llvm-ir) below).
```

Comment:
Also, the LLVM bindings using Cgo are in general not very idiomatic Go and somewhat inconsistent. But that is perhaps too much for this blog post.

Reply:
> Also, the LLVM bindings using Cgo are in general not very idiomatic Go and somewhat inconsistent. But that is perhaps too much for this blog post.

I mentioned this above, *The official LLVM bindings for Go uses Cgo* and in the rant of how LLVM values are modelled. ^^